### PR TITLE
Use TypeToken.getParameterized for all parameterized fields

### PR DESCRIPTION
**Pull request by benkay**
This eliminates the remaining reflection when creating TypeAdapters.

In a test app on a Galaxy J1, it reduces the time taken to initialize 10 `TypeAdapter`s for classes each containing 10 `List<String>` fields from 750 - 800ms to ~30ms

I also enabled custom field adapter support for parameterized generic fields. Not sure it's useful, but it simplified the code.

### CODE REVIEW

File: auto-value-gson/src/main/java/com/ryanharter/auto/value/gson/AutoValueGsonExtension.java [hunk 1]
```diff
@@ -330,7 +330,6 @@ public TypeSpec createTypeAdapter(Context context, ClassName className, ClassNam
             .addParameter(gsonParam);
 
     if (!typeParams.isEmpty()) {
-      ImmutableMap<Property, FieldSpec> parameterizedAdapters = filterParameterizedAdapters(adapters, true);
```
Comment (JakeWharton) [inline auto-value-gson/src/main/java/com/ryanharter/auto/value/gson/AutoValueGsonExtension.java]: nit: delete empty line

**Comment by thevery** [comment]
Hm, looks like this doesn't work with proguard nicely:

Warning: com.yandex.mail.react.entity.$AutoValue_Fields$GsonTypeAdapter: can't find referenced method 'com.google.gson.reflect.TypeToken getParameterized(java.lang.reflect.Type,java.lang.reflect.Type[])' in program class com.google.gson.reflect.TypeToken

**Comment by thevery** [comment]
upd: not an auto-value-gson problem but rather gson: TypeToken.getParameterized(List.class, Recipient.class) was added in gson 2.8.0

**Comment by thevery** [comment]
upd2: my bad, forgot to update explicit gson dependency to 2.8.

**Pull request merged**